### ISSUE TITLE
New Rule: disallow unsafe use of innerHTML and insertAdjacentHTML

### ISSUE THREAD

**Issue by mozfreddyb** [eslint/eslint#2496]
I am currently writing a security rule for eslint (initial PR in #2495).

The idea is to disallow the following:
```js
foo.innerHTML = input.value;
bar.innerHTML = "<a href='"+url+"'>About</a>";
```

We want to use eslint to enforce a mandatory escaping function. For the escaping function to be sufficiently secure, we need it to get an ES6 template string, to distinguish HTML templates from input variables. That means, what we want to allow is this:

```js
bar.innerHTML = escapeHTML`<a href='${url}'>About</a>`;
```

We also want to allow trivial cases, where security is not in the hand of the escape function but the code reviewer (e.g., using any kind hard coded of literal):
```js
foo.innerHTML = "<a href='/about'>About</a>";
```

I'd be interested to discuss if this of any value to eslint itself. If not I'm happy to maintain this as an eslint plugin.

**Comment by nzakas** [comment]
I'm not sure I understand the conditions under which you want to warn. Basically anytime `innerHTML` is used with anything other than a single string literal?

**Comment by gcochard** [comment]
This seems to be a function of DOM security, and not something I'd vote for as a built-in rule. Eslint has many uses outside of the browser environment. Definitely a very useful plugin though.

**Comment by mozfreddyb** [comment]
Anything warns, except than a literal, a templateliteral or a tagged template string where the function is one of the whitelisted function names.

**Comment by nzakas** [comment]
Yeah, I think this might be better in a plugin as I'm not sure it's general enough for inclusion in the core. For instance, a single string is also not guaranteed to be safe, so would we then need to support multiple options for which things are allowed or not?

**Comment by mozfreddyb** [comment]
A single string is also a literal node, it would be covered by my previous statement and it is allowed.

**Comment by nzakas** [comment]
Closing. My opinion remains that this is best as a plugin.

**Comment by mozfreddyb** [comment]
Alright, thanks for taking this into consideration.
This is available as `eslint-plugin-no-unsafe-innerhtml` on npm and github now.
